### PR TITLE
Convert savefile previews to use aynchronous client messaging.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR converts retrieving save file preview from the server from using the synchronous client messaging to asynchronous.

The problems are the same as the problems with synchronous requests of object IDs, client glitches up to and including crashes.

### CODE REVIEW

File: UI/SaveFileDialog.h [hunk 1]
```diff
@@ -38,6 +35,9 @@ class SaveFileDialog : public CUIWnd {
     void ModalInit() override;
 
     void KeyPress(GG::Key key, std::uint32_t key_code_point, GG::Flags<GG::ModKey> mod_keys) override;
+
+    /** Update the previews with \p preview_information*/
```
Comment (dbenage-cx) [inline UI/SaveFileDialog.h]: preview_info

**Comment by geoffthemedio** [comment]
When trying to load a saved multiplayer game, with this pull request applied, I don't get anything listed in the file dialog. The log says the client received no previews, but there are numerous .mps files it could have found...
no_previews

Edit: looks like master doesn't find them either, though...

Edit2: added display of the number of subdirs in the receive preview info, and they are shown, but just not shown, so it's a client-side display issue for MP save files, seemingly:
````
@@ -794,7 +794,8 @@ void HumanClientApp::RequestSavePreviews(const std::string& directory, PreviewIn
     const auto response = m_networking->SendSynchronousMessage(RequestSavePreviewsMessage(generic_directory));
     if (response && response->Type() == Message::DISPATCH_SAVE_PREVIEWS) {
         ExtractDispatchSavePreviewsMessageData(*response, previews);
-        DebugLogger() << "HumanClientApp::RequestSavePreviews Got " << previews.previews.size() << " previews.";
+        DebugLogger() << "HumanClientApp::RequestSavePreviews Got " << previews.previews.size()
+                      << " previews and " << previews.subdirectories.size() << " subdirectories";
     } else {
         ErrorLogger() << "HumanClientApp::RequestSavePreviews: Wrong response type from server: " << response->Type();
     }
````

**Comment by LGM-Doyle** [comment]
As you observed that was the existing behavior.

I refactored both the server and client so that server side save work more like client side saves.

On the server side, I added a server save dir option, changed `UpdateSavePreviews()` to accept only relative directories and only return directories and files that are within the server save dir.  

On the client side, I removed the remote server dir drop down. The server dir drop down was where the directories were supposed to be shown and changed, but it was different from the client side save UI and non-obvious.  I made both server side and client side saves use the same display except server directories are prefixed with "SERVER/".

File: UI/SaveFileDialog.h (hunk 1 above)
Comment (LGM-Doyle) [reply UI/SaveFileDialog.h]: Fixed.

**Comment by geoffthemedio** [comment]
Better, although non-Latin characters seem to have issues with display, and I can't navigate into the directory shown with ??? in the name in the GUI. It's actually `asdfШит`.
server_saves

**Comment by LGM-Doyle** [comment]
I can't replicate either of those problems, the display or the traversal of directories with non-Latin character names.

I think the problem is Windows uses UTF-16 characters for directory names. Someone with access to Windows could add an `#ifdef` to convert the Windows directory names to/from UTF-8 which is what `GG` is using for display and `boost::filesystem` is using to list the directories.

File: server/ServerApp.cpp [hunk 2]
```diff
@@ -875,32 +859,96 @@ void ServerApp::LoadSPGameInit(const std::vector<PlayerSaveGameData>& player_sav
     LoadGameInit(player_save_game_data, player_id_to_save_game_data_index, server_save_game_data);
 }
 
+namespace {
+        /** Check that \p path is a file or directory in the server save
+        directory. */
+    bool IsInServerSaveDir(const fs::path& path) {
+
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: extra blank line?

File: server/ServerApp.cpp [hunk 3]
```diff
@@ -875,32 +859,96 @@ void ServerApp::LoadSPGameInit(const std::vector<PlayerSaveGameData>& player_sav
...
+        return std::equal(canon_server.begin(), canon_server.end(), path.begin());
+    }
+
+    /// Generates information on the subdirectories of \p directory
+    std::vector<std::string> ListSaveSubdirectories(const fs::path& directory) {
+        std::vector<std::string> list;
+        if (!fs::is_directory(directory))
+            return list;
+
+        std::string server_dir_str = fs::canonical(GetServerSaveDir()).generic_string();
```
Comment (geoffthemedio) [inline server/ServerApp.cpp]: Could probably call `PathString` here, and `FilenameToPath` whenever taking a string received in a message from the client that needs to be converted back into a path. This way, all the pathes are sent to and from the client as UTF-8, and only converted back to UTF-16 when creating a `path` from a string on Windows.

Comment (LGM-Doyle) [reply server/ServerApp.cpp]: Thanks for looking into that issue.

I've made the changes.  It might now also work on Windows

**Comment by geoffthemedio** [comment]
I see the directory name now, but clicking it still doesn't work, and when I do, I very briefly see a grabled version appear in the upper path box. I think pathes are still being sent or converted incorrectly in this process...

dirprob

**Comment by LGM-Doyle** [comment]
Here is another attempt, converting fragments of filenames to/from UTF8.

**Comment by geoffthemedio** [comment]
same result as last reported

In freeoriond.log:
````
09:01:28.693429 [debug] server : ServerApp.cpp:928 : ServerApp::UpdateSavePreviews: Preview request for sub directory: ./asdfÐ¨Ð¸Ñ‚
09:01:28.693429 [error] server : ServerApp.cpp:935 : ServerApp::UpdateSavePreviews: Tried to load previews from ./asdfÐ¨Ð¸Ñ‚ which is outside the allowed save directory. Defaulted to the save directory, "C:\Users\Geoff\AppData\Roaming\FreeOrion\save"
09:01:28.696430 [debug] server : ServerApp.cpp:949 : ServerApp::UpdateSavePreviews: Sending 0 previews in response.
````
In freeorion.log:
````
09:01:27.878706 [debug] client : HumanClientApp.cpp:792 : HumanClientApp::RequestSavePreviews Requesting previews for ./asdfÐ¨Ð¸Ñ‚
09:01:27.896827 [debug] client : HumanClientApp.cpp:983 : HumanClientApp::RequestSavePreviews Got 0 previews.
09:01:27.899833 [debug] client : SaveFileDialog.cpp:597 : SaveFileDialog::FilterColumns: Visible columns: 5
09:01:28.692441 [debug] client : SaveFileDialog.cpp:799 : SaveFileDialog::Confirm: Confirming
09:01:28.692441 [debug] client : SaveFileDialog.cpp:815 : choice: SERVER\asdfШит valid utf-8: true
09:01:28.692441 [debug] client : SaveFileDialog.cpp:1027 : SaveFileDialog::GetDirPath retval: ./. valid UTF-8: true
09:01:28.692441 [debug] client : SaveFileDialog.cpp:818 : current dir PathString: ./. valid utf-8: true
09:01:28.692441 [debug] client : SaveFileDialog.cpp:821 : chosen_full_path PathString: ./.\SERVER\asdfШит valid utf-8: true
09:01:28.692441 [debug] client : SaveFileDialog.cpp:822 : chosen_full_path is directory? : false
09:01:28.693429 [debug] client : SaveFileDialog.cpp:1027 : SaveFileDialog::GetDirPath retval: ./asdfÐ¨Ð¸Ñ‚ valid UTF-8: true
09:01:28.693429 [debug] client : HumanClientApp.cpp:792 : HumanClientApp::RequestSavePreviews Requesting previews for ./asdfÐ¨Ð¸Ñ‚
09:01:28.710704 [debug] client : HumanClientApp.cpp:983 : HumanClientApp::RequestSavePreviews Got 0 previews.
09:01:28.713694 [debug] client : SaveFileDialog.cpp:597 : SaveFileDialog::FilterColumns: Visible columns: 5
09:01:31.365891 [debug] client : HumanClientApp.cpp:1042 : HumanClientApp::HandleFocusChange(Lost Focus)
````

**Comment by LGM-Doyle** [comment]
Thanks for the logs.

I have wrapped the results of `GetDirPath()` in `GenericPathString()` which should convert both the displayed and sent to server path components to UTF-8.

**Comment by geoffthemedio** [comment]
Seemingly no change in behaviour.
````
22:07:18.744296 [debug] client : SaveFileDialog.cpp:815 : choice: SERVER\asdfШит valid utf-8: true
22:07:18.744296 [debug] client : SaveFileDialog.cpp:1027 : SaveFileDialog::GetDirPath retval: ./. valid UTF-8: true
22:07:18.744296 [debug] client : SaveFileDialog.cpp:818 : current dir PathString: ./. valid utf-8: true
22:07:18.744296 [debug] client : SaveFileDialog.cpp:821 : chosen_full_path PathString: ./.\SERVER\asdfШит valid utf-8: true
22:07:18.744296 [debug] client : SaveFileDialog.cpp:822 : chosen_full_path is directory? : false
22:07:18.744296 [debug] client : SaveFileDialog.cpp:1027 : SaveFileDialog::GetDirPath retval: ./asdfÃÂ¨ÃÂ¸Ã‘â€š valid UTF-8: true
22:07:18.744296 [debug] client : HumanClientApp.cpp:792 : HumanClientApp::RequestSavePreviews Requesting previews for ./asdfÃÂ¨ÃÂ¸Ã‘â€š
22:07:18.761340 [debug] client : HumanClientApp.cpp:983 : HumanClientApp::RequestSavePreviews Got 0 previews.
22:07:18.763341 [debug] client : SaveFileDialog.cpp:597 : SaveFileDialog::FilterColumns: Visible columns: 5
22:07:20.426682 [debug] client : SaveFileDialog.cpp:896 : SaveFileDialog::Cancel: Dialog Canceled
````
````
22:07:14.759873 [debug] server : ServerApp.cpp:928 : ServerApp::UpdateSavePreviews: Preview request for sub directory: ./
22:07:14.762864 [debug] server : ServerApp.cpp:949 : ServerApp::UpdateSavePreviews: Sending 0 previews in response.
22:07:16.652080 [debug] server : ServerApp.cpp:928 : ServerApp::UpdateSavePreviews: Preview request for sub directory: ./asdfÃÂ¨ÃÂ¸Ã‘â€š
22:07:16.653093 [error] server : ServerApp.cpp:935 : ServerApp::UpdateSavePreviews: Tried to load previews from ./asdfÃÂ¨ÃÂ¸Ã‘â€š which is outside the allowed save directory. Defaulted to the save directory, "C:\Users\Geoff\AppData\Roaming\FreeOrion\save"
22:07:16.656082 [debug] server : ServerApp.cpp:949 : ServerApp::UpdateSavePreviews: Sending 0 previews in response.
22:07:18.744296 [debug] server : ServerApp.cpp:928 : ServerApp::UpdateSavePreviews: Preview request for sub directory: ./asdfÃÂ¨ÃÂ¸Ã‘â€š
22:07:18.745287 [error] server : ServerApp.cpp:935 : ServerApp::UpdateSavePreviews: Tried to load previews from ./asdfÃÂ¨ÃÂ¸Ã‘â€š which is outside the allowed save directory. Defaulted to the save directory, "C:\Users\Geoff\AppData\Roaming\FreeOrion\save"
22:07:18.749299 [debug] server : ServerApp.cpp:949 : ServerApp::UpdateSavePreviews: Sending 0 previews in response.
````

File: server/ServerApp.cpp [hunk 4]
```diff
@@ -875,32 +859,95 @@ void ServerApp::LoadSPGameInit(const std::vector<PlayerSaveGameData>& player_sav
...
+        auto canon_path = fs::canonical(path);
+
+        // Paths shorter than the server save dir are not in the server dir
+        auto save_dir_length = std::distance(canon_server.begin(), canon_server.end());
+        auto path_length = std::distance(canon_path.begin(), canon_path.end());
+        if (path_length < save_dir_length)
+            return false;
+
+        // Check that the whole server save dir path matches the test path
+        return std::equal(canon_server.begin(), canon_server.end(), path.begin());
```
Comment (dbenage-cx) [inline server/ServerApp.cpp]: Should this be `canon_path` instead of `path`?

**Comment by LGM-Doyle** [comment]
Thanks again for the logs.

Another try. I think the problem with the Windows filenames was that they were being doubly converted on the server and then again on the client.  The additional commits restrict it to a single conversion.

This also makes the use of the `ServerSaveDir()` more consistent.

**Pull request merged**